{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1a104969",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "import gffutils\n",
    "import pandas as pd\n",
    "import sqlite3\n",
    "\n",
    "in_dir = \"input\"\n",
    "out_dir = \"output\"\n",
    "Path(out_dir).mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "075b843a",
   "metadata": {},
   "source": [
    "## Create GFF databases and read in features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "30604c64",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_db(path, db_filename, **kwargs):\n",
    "    \"\"\"\n",
    "    Create/connect db for given gff file path\n",
    "    \"\"\"\n",
    "    if not os.path.isfile(os.path.join(out_dir, db_filename)):\n",
    "        db = gffutils.create_db(path, os.path.join(out_dir, db_filename), force=True, **kwargs)\n",
    "    else:\n",
    "        db = sqlite3.connect(os.path.join(out_dir, db_filename), check_same_thread=False)\n",
    "        db = gffutils.FeatureDB(db)\n",
    "    return db\n",
    "\n",
    "# WormBase release WS283 canonical three_prime_UTR for C. elegans chromosome I\n",
    "control_fn = os.path.join(in_dir, \"WormBase.I.three_prime_UTR.gff3\")\n",
    "db_control = create_db(control_fn, \"control.db\")\n",
    "control = list(db_control.all_features())\n",
    "\n",
    "# Output from peaks2utr run\n",
    "test_fn = os.path.join(in_dir, \"peaks2utr_output.gff3\")\n",
    "db_test = create_db(test_fn, \"test.db\")\n",
    "test = list(db_test.all_features())\n",
    "\n",
    "# WormBase release WS283 canonical mRNA transcripts for C. elegans chromosome I\n",
    "genes_fn = os.path.join(in_dir, \"WormBase.I.mRNA.gff3\")\n",
    "db_genes = create_db(genes_fn, \"genes.db\")\n",
    "genes = list(db_genes.all_features())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcbcdfd2",
   "metadata": {},
   "source": [
    "## Prepare test dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf250ec8",
   "metadata": {},
   "source": [
    "Create a dataframe of peaks2utr test UTRs with a \"gene id\" index."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a45d2844",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WBGene00022277</th>\n",
       "      <td>4114</td>\n",
       "      <td>4220</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00022276</th>\n",
       "      <td>16586</td>\n",
       "      <td>16906</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00022278</th>\n",
       "      <td>17482</td>\n",
       "      <td>17910</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00022279</th>\n",
       "      <td>27588</td>\n",
       "      <td>28279</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00044345</th>\n",
       "      <td>49417</td>\n",
       "      <td>49827</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000881</th>\n",
       "      <td>15040405</td>\n",
       "      <td>15040608</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00009285</th>\n",
       "      <td>15041428</td>\n",
       "      <td>15041884</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00009286</th>\n",
       "      <td>15050795</td>\n",
       "      <td>15051145</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00009287</th>\n",
       "      <td>15053619</td>\n",
       "      <td>15053673</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00009288</th>\n",
       "      <td>15053603</td>\n",
       "      <td>15053793</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2278 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   start       end strand\n",
       "gene                                     \n",
       "WBGene00022277      4114      4220      -\n",
       "WBGene00022276     16586     16906      +\n",
       "WBGene00022278     17482     17910      -\n",
       "WBGene00022279     27588     28279      -\n",
       "WBGene00044345     49417     49827      +\n",
       "...                  ...       ...    ...\n",
       "WBGene00000881  15040405  15040608      -\n",
       "WBGene00009285  15041428  15041884      -\n",
       "WBGene00009286  15050795  15051145      -\n",
       "WBGene00009287  15053619  15053673      +\n",
       "WBGene00009288  15053603  15053793      -\n",
       "\n",
       "[2278 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tdf = pd.DataFrame([vars(t) for t in test], columns=[\"start\", \"end\", \"strand\", \"attributes\"])\n",
    "tdf['name'] = tdf[\"attributes\"].apply(lambda x: x.get(\"sequence_name\")[0])\n",
    "tdf[\"gene\"] = tdf[\"attributes\"].apply(lambda x: x.get(\"Name\")[0])\n",
    "test_utrs = tdf.drop([\"name\", \"attributes\"], axis=1).set_index(\"gene\")\n",
    "test_utrs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05519a99",
   "metadata": {},
   "source": [
    "## Prepare control dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba3fe4a0",
   "metadata": {},
   "source": [
    "C. elegans contains alternative mRNA transcripts for each gene, and the control UTRs refer only to these mRNA transcript \"names\" in their attributes. In this instance, peaks2utr has been configured to call UTRs on the parent gene feature itself, rather than the child transcripts. Therefore, it will be helpful to create a sequence name -> gene id mapping dataframe to allow comparisions between the two datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "76d7a055",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Y74C9A.3.1</td>\n",
       "      <td>WBGene00022277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Y74C9A.2a.3</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Y74C9A.2a.1</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Y74C9A.2a.2</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Y74C9A.2b.1</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4698</th>\n",
       "      <td>F31C3.1.1</td>\n",
       "      <td>WBGene00000881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4699</th>\n",
       "      <td>F31C3.3.1</td>\n",
       "      <td>WBGene00009285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4700</th>\n",
       "      <td>F31C3.4.1</td>\n",
       "      <td>WBGene00009286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4701</th>\n",
       "      <td>F31C3.5.1</td>\n",
       "      <td>WBGene00009287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4702</th>\n",
       "      <td>F31C3.6a.1</td>\n",
       "      <td>WBGene00009288</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4703 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             name            gene\n",
       "0      Y74C9A.3.1  WBGene00022277\n",
       "1     Y74C9A.2a.3  WBGene00022276\n",
       "2     Y74C9A.2a.1  WBGene00022276\n",
       "3     Y74C9A.2a.2  WBGene00022276\n",
       "4     Y74C9A.2b.1  WBGene00022276\n",
       "...           ...             ...\n",
       "4698    F31C3.1.1  WBGene00000881\n",
       "4699    F31C3.3.1  WBGene00009285\n",
       "4700    F31C3.4.1  WBGene00009286\n",
       "4701    F31C3.5.1  WBGene00009287\n",
       "4702   F31C3.6a.1  WBGene00009288\n",
       "\n",
       "[4703 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdf = pd.DataFrame([vars(c) for c in genes], columns=[\"featuretype\", \"start\", \"end\", \"strand\", \"attributes\"])\n",
    "gdf = gdf[gdf[\"featuretype\"] == \"mRNA\"]\n",
    "gdf[\"gene\"] = gdf[\"attributes\"].apply(lambda x: x.get(\"Parent\")[0].split(\":\")[-1])\n",
    "gdf[\"name\"] = gdf[\"attributes\"].apply(lambda x: x.get(\"Name\")[0])\n",
    "gdf = gdf[[\"name\", \"gene\"]]\n",
    "gdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a524a26b",
   "metadata": {},
   "source": [
    "Create a dataframe of canonical WormBase control UTRs with a \"name\" column consisting of the parent gene's name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "60b0dc8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4116</td>\n",
       "      <td>4220</td>\n",
       "      <td>-</td>\n",
       "      <td>Y74C9A.3.1</td>\n",
       "      <td>WBGene00022277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16586</td>\n",
       "      <td>16833</td>\n",
       "      <td>+</td>\n",
       "      <td>Y74C9A.2a.2</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16586</td>\n",
       "      <td>16837</td>\n",
       "      <td>+</td>\n",
       "      <td>Y74C9A.2a.1</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16702</td>\n",
       "      <td>16793</td>\n",
       "      <td>+</td>\n",
       "      <td>Y74C9A.2a.3</td>\n",
       "      <td>WBGene00022276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17484</td>\n",
       "      <td>17910</td>\n",
       "      <td>-</td>\n",
       "      <td>Y74C9A.4a.1</td>\n",
       "      <td>WBGene00022278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4192</th>\n",
       "      <td>15040474</td>\n",
       "      <td>15040608</td>\n",
       "      <td>-</td>\n",
       "      <td>F31C3.1.1</td>\n",
       "      <td>WBGene00000881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4193</th>\n",
       "      <td>15041433</td>\n",
       "      <td>15041884</td>\n",
       "      <td>-</td>\n",
       "      <td>F31C3.3.1</td>\n",
       "      <td>WBGene00009285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4194</th>\n",
       "      <td>15050797</td>\n",
       "      <td>15051145</td>\n",
       "      <td>-</td>\n",
       "      <td>F31C3.4.1</td>\n",
       "      <td>WBGene00009286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4195</th>\n",
       "      <td>15053608</td>\n",
       "      <td>15053793</td>\n",
       "      <td>-</td>\n",
       "      <td>F31C3.6a.1</td>\n",
       "      <td>WBGene00009288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4196</th>\n",
       "      <td>15053619</td>\n",
       "      <td>15053672</td>\n",
       "      <td>+</td>\n",
       "      <td>F31C3.5.1</td>\n",
       "      <td>WBGene00009287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4197 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         start       end strand         name            gene\n",
       "0         4116      4220      -   Y74C9A.3.1  WBGene00022277\n",
       "1        16586     16833      +  Y74C9A.2a.2  WBGene00022276\n",
       "2        16586     16837      +  Y74C9A.2a.1  WBGene00022276\n",
       "3        16702     16793      +  Y74C9A.2a.3  WBGene00022276\n",
       "4        17484     17910      -  Y74C9A.4a.1  WBGene00022278\n",
       "...        ...       ...    ...          ...             ...\n",
       "4192  15040474  15040608      -    F31C3.1.1  WBGene00000881\n",
       "4193  15041433  15041884      -    F31C3.3.1  WBGene00009285\n",
       "4194  15050797  15051145      -    F31C3.4.1  WBGene00009286\n",
       "4195  15053608  15053793      -   F31C3.6a.1  WBGene00009288\n",
       "4196  15053619  15053672      +    F31C3.5.1  WBGene00009287\n",
       "\n",
       "[4197 rows x 5 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdf = pd.DataFrame([vars(c) for c in control], columns=[\"start\", \"end\", \"source\", \"strand\", \"attributes\"])\n",
    "cdf = cdf[cdf.source == \"WormBase\"]\n",
    "cdf['name'] = cdf[\"attributes\"].apply(lambda x: x.get(\"Parent\")[0].split(\":\")[-1])\n",
    "cdf = cdf.drop(['attributes', \"source\"], axis=1)\n",
    "cdf = pd.merge(cdf, gdf, on=\"name\")\n",
    "cdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1c5da8c",
   "metadata": {},
   "source": [
    "As peaks2utr has used the --override-utr flag, we know that the start/end base will match the start/end base of the 3' UTR for the innermost mRNA transcript (on the forward/reverse strand, respectively). Therefore, we can aggregate each gene by taking the minimum start base and maximum end base to capture the full extent of the UTR for that gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "35a5f8b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WBGene00000001</th>\n",
       "      <td>+</td>\n",
       "      <td>5109948</td>\n",
       "      <td>5110183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000006</th>\n",
       "      <td>-</td>\n",
       "      <td>2571740</td>\n",
       "      <td>2571987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000010</th>\n",
       "      <td>-</td>\n",
       "      <td>11388682</td>\n",
       "      <td>11389140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000012</th>\n",
       "      <td>-</td>\n",
       "      <td>3886070</td>\n",
       "      <td>3886110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000020</th>\n",
       "      <td>-</td>\n",
       "      <td>2255921</td>\n",
       "      <td>2256151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00304996</th>\n",
       "      <td>+</td>\n",
       "      <td>5341995</td>\n",
       "      <td>5342045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00305018</th>\n",
       "      <td>-</td>\n",
       "      <td>9588662</td>\n",
       "      <td>9588719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00306080</th>\n",
       "      <td>-</td>\n",
       "      <td>2939477</td>\n",
       "      <td>2939518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00306081</th>\n",
       "      <td>+</td>\n",
       "      <td>6872119</td>\n",
       "      <td>6872142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00306121</th>\n",
       "      <td>-</td>\n",
       "      <td>231578</td>\n",
       "      <td>231691</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2593 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               strand     start       end\n",
       "gene                                     \n",
       "WBGene00000001      +   5109948   5110183\n",
       "WBGene00000006      -   2571740   2571987\n",
       "WBGene00000010      -  11388682  11389140\n",
       "WBGene00000012      -   3886070   3886110\n",
       "WBGene00000020      -   2255921   2256151\n",
       "...               ...       ...       ...\n",
       "WBGene00304996      +   5341995   5342045\n",
       "WBGene00305018      -   9588662   9588719\n",
       "WBGene00306080      -   2939477   2939518\n",
       "WBGene00306081      +   6872119   6872142\n",
       "WBGene00306121      -    231578    231691\n",
       "\n",
       "[2593 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_utrs = cdf.groupby([\"gene\", \"strand\"]).agg({\"start\": min, \"end\": max}).reset_index(\"strand\")\n",
    "control_utrs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b8ed9f8",
   "metadata": {},
   "source": [
    "# Compare the two datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "262bfe01",
   "metadata": {},
   "source": [
    "Now that we have two comparable datasets with paired UTRs, we can calculate some statistical differences. We determine length of each UTR in both dataframes, and determine differences of test UTR length to true UTR length."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5b72a67b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>len</th>\n",
       "      <th>difference</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WBGene00022277</th>\n",
       "      <td>4114</td>\n",
       "      <td>4220</td>\n",
       "      <td>-</td>\n",
       "      <td>106</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00022276</th>\n",
       "      <td>16586</td>\n",
       "      <td>16906</td>\n",
       "      <td>+</td>\n",
       "      <td>320</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00022278</th>\n",
       "      <td>17482</td>\n",
       "      <td>17910</td>\n",
       "      <td>-</td>\n",
       "      <td>428</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00022279</th>\n",
       "      <td>27588</td>\n",
       "      <td>28279</td>\n",
       "      <td>-</td>\n",
       "      <td>691</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00044345</th>\n",
       "      <td>49417</td>\n",
       "      <td>49827</td>\n",
       "      <td>+</td>\n",
       "      <td>410</td>\n",
       "      <td>-33.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000881</th>\n",
       "      <td>15040405</td>\n",
       "      <td>15040608</td>\n",
       "      <td>-</td>\n",
       "      <td>203</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00009285</th>\n",
       "      <td>15041428</td>\n",
       "      <td>15041884</td>\n",
       "      <td>-</td>\n",
       "      <td>456</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00009286</th>\n",
       "      <td>15050795</td>\n",
       "      <td>15051145</td>\n",
       "      <td>-</td>\n",
       "      <td>350</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00009287</th>\n",
       "      <td>15053619</td>\n",
       "      <td>15053673</td>\n",
       "      <td>+</td>\n",
       "      <td>54</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00009288</th>\n",
       "      <td>15053603</td>\n",
       "      <td>15053793</td>\n",
       "      <td>-</td>\n",
       "      <td>190</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2278 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   start       end strand  len  difference\n",
       "gene                                                      \n",
       "WBGene00022277      4114      4220      -  106         2.0\n",
       "WBGene00022276     16586     16906      +  320        69.0\n",
       "WBGene00022278     17482     17910      -  428         2.0\n",
       "WBGene00022279     27588     28279      -  691         1.0\n",
       "WBGene00044345     49417     49827      +  410       -33.0\n",
       "...                  ...       ...    ...  ...         ...\n",
       "WBGene00000881  15040405  15040608      -  203        69.0\n",
       "WBGene00009285  15041428  15041884      -  456         5.0\n",
       "WBGene00009286  15050795  15051145      -  350         2.0\n",
       "WBGene00009287  15053619  15053673      +   54         1.0\n",
       "WBGene00009288  15053603  15053793      -  190         5.0\n",
       "\n",
       "[2278 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_utrs[\"len\"] = abs(control_utrs[\"end\"] - control_utrs[\"start\"])\n",
    "test_utrs[\"len\"] = abs(test_utrs[\"end\"] - test_utrs[\"start\"])\n",
    "test_utrs[\"difference\"] = (test_utrs[\"len\"] - control_utrs[\"len\"])\n",
    "test_utrs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "522098d2",
   "metadata": {},
   "source": [
    "## Supplementary Table 2 statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "25272924",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Canonical utrs that are missed by peaks2utr: 451 (17.39%)\n",
      "peaks2utr utrs that match true utr to within 50 bases: 1578 (60.86%)\n",
      "peaks2utr utrs that reduce true utr by >50 bases: 117 (4.51%)\n",
      "peaks2utr utrs that extend true utr by between 50 and 200 bases: 211 (8.14%)\n",
      "peaks2utr utrs that extend true utr by >200 bases: 236 (9.10%)\n",
      "peaks2utr utrs called where no true utr existed previously: 136\n",
      "Total peaks2utr UTRs: 2278\n"
     ]
    }
   ],
   "source": [
    "total_test = len(test_utrs)\n",
    "total_control = len(control_utrs)\n",
    "\n",
    "genes_missed_utr = control_utrs.index.difference(test_utrs.index)\n",
    "num_missed_utrs = len(genes_missed_utr)\n",
    "genes_no_utr = set(gdf[\"gene\"]).difference(control_utrs.index)\n",
    "novel_utr_genes = genes_no_utr.intersection(test_utrs.index)\n",
    "num_novel_utrs = len(novel_utr_genes)\n",
    "\n",
    "BP1 = 50\n",
    "BP2 = 200\n",
    "genes_matched_utr =  test_utrs[abs(test_utrs[\"difference\"]) <= BP1].index\n",
    "num_matched = len(genes_matched_utr)\n",
    "num_reduced = len(test_utrs[test_utrs['difference'] < -BP1])\n",
    "num_extended1 = len(test_utrs[(test_utrs['difference'] > BP1) & (test_utrs['difference'] <= BP2)])\n",
    "num_extended2 = len(test_utrs[test_utrs['difference'] > BP2])\n",
    "\n",
    "print(\"Canonical utrs that are missed by peaks2utr: {} ({:.2f}%)\".format(num_missed_utrs, 100*num_missed_utrs/total_control))\n",
    "print(\"peaks2utr utrs that match true utr to within {} bases: {} ({:.2f}%)\".format(BP1, num_matched, 100*num_matched/total_control))\n",
    "print(\"peaks2utr utrs that reduce true utr by >{} bases: {} ({:.2f}%)\".format(BP1, num_reduced, 100*num_reduced/total_control))\n",
    "print(\"peaks2utr utrs that extend true utr by between {} and {} bases: {} ({:.2f}%)\".format(BP1, BP2, num_extended1, 100*num_extended1/total_control))\n",
    "print(\"peaks2utr utrs that extend true utr by >{} bases: {} ({:.2f}%)\".format(BP2, num_extended2, 100*num_extended2/total_control))\n",
    "print(\"peaks2utr utrs called where no true utr existed previously: {}\".format(num_novel_utrs))\n",
    "print(\"Total peaks2utr UTRs: {}\".format(total_test))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec34014f",
   "metadata": {},
   "source": [
    "Save lists of gene ids with novel, matched and missed 3' UTRs called by peaks2utr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c1f7b246",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(out_dir, \"novel_utr_genes_peaks2utr.txt\"), \"w\") as f:\n",
    "    for g in sorted(list(novel_utr_genes)):\n",
    "        f.write(g + \"\\n\")\n",
    "\n",
    "with open(os.path.join(out_dir, \"matched_utr_genes_peaks2utr.txt\"), \"w\") as f:\n",
    "    for g in sorted(list(genes_matched_utr)):\n",
    "        f.write(g + \"\\n\")\n",
    "\n",
    "with open(os.path.join(out_dir, \"missed_utr_genes_peaks2utr.txt\"), \"w\") as f:\n",
    "    for g in sorted(list(genes_missed_utr)):\n",
    "        f.write(g + \"\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
